---
title: "Computing a PGS using RápidoPGS and GWAS catalog"
author: "Guillermo Reales"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Computing a PGS using RápidoPGS and GWAS catalog}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}

---

# Introduction

RápidoPGS is a tool to quickly compute polygenic scores (PGS) from GWAS summary statistics datasets from either case-control (eg. asthma) or quantitative traits (eg. height and BMI).

Input GWAS summary statistics datasets should have a minimum set of columns, 
with these names, but in any order:


---
title: "Computing a PGS using RápidoPGS and GWAS catalog"
author: "Guillermo Reales"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Computing a PGS using RápidoPGS and GWAS catalog}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}

---

# Introduction

RápidoPGS is a tool to quickly compute polygenic scores (PGS) from GWAS summary statistics datasets from either case-control (eg. asthma) or quantitative traits (eg. height and BMI).

Input GWAS summary statistics datasets should have a minimum set of columns, 
with these names, but in any order:

---
title: "Computing a PGS using RápidoPGS and GWAS catalog"
author: "Guillermo Reales"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Computing a PGS using RápidoPGS and GWAS catalog}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}

---

# Introduction

RápidoPGS is a tool to quickly compute polygenic scores (PGS) from GWAS summary statistics datasets from either case-control (eg. asthma) or quantitative traits (eg. height and BMI).

Input GWAS summary statistics datasets should have a minimum set of columns, 
with these names, but in any order:


---
title: "Computing a PGS using RápidoPGS and GWAS catalog"
author: "Guillermo Reales"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Computing a PGS using RápidoPGS and GWAS catalog}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}

---

# Introduction

RápidoPGS is a tool to quickly compute polygenic scores (PGS) from GWAS summary statistics datasets from either case-control (eg. asthma) or quantitative traits (eg. height and BMI). It also provides a wrapper to retrieve public, harmonised data from GWAS catalog, which use we'll demonstrate in this vignette, too.

Input GWAS summary statistics datasets should have a minimum set of columns, 
with these names, but in any order:

* CHR: Chromosome 
* BP: Base position (in GRCh37/hg19 or GRCh38/hg38). If using hg38, use build = "hg38" in parameters.
* SNPID: rsids, or SNP identifiers. If not available, they can be anything (eg. CHR_BP).
* REF: Reference, or non-effect allele
* ALT: Alternative, or effect allele, the one \eqn{\beta} refers to
* BETA: $\beta$ (or log(OR)), or effect sizes.
* SE: Standard error of $\beta$
* P: P-values for the association test
* ALT_FREQ: Minor/ALT allele frequency in the tested population, or in a close population from a reference panel. No problem it flipped.


# Installation of RápidoPGS

RápidoPGS is currently available in GitHub, so we can install it using the following code (note: If you don't have `devtools` installed, please install it first:

```{r eval=FALSE}
library(devtools)
install_github("GRealesM/RapidoPGS")

# If you might need to install GenomicRanges, you can use the following code
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("GenomicRanges")
```

# Setup

Once installed, let's load it. This will automatically load all required dependencies.

```{r}
library(RapidoPGS)
```

# Downloading data from GWAS catalog

[GWAS catalog](https://www.ebi.ac.uk/gwas/) is an outstanding GWAS summary statistics data source, as it not only puts together tons of public dataset, but it also uses a semi-automatic pipeline to apply quality control and liftover (a.k.a. harmonise) those datasets, thus helping overcome the problem of having GWAS sumstats in so many different formats and builds.

In this vignette, we'll use GWAS catalog to obtain an example dataset. For this vignette we'll use a Breast cancer (BRCA) dataset from [Michailidou et al., 2017](https://www.nature.com/articles/nature24284), which is one that we used in our manuscript. This GWAS was performed on 119078 controls and 137045 cases of Breast cancer.

We use it's PubMed ID (29059683), and set `hm_only = FALSE` to get all columns.

It's a big file, so it may take a while to download. 
Note that in this particular case we chose a study that contained a single dataset. In other cases there may be more than one. In that case `gwascat.download` will prompt you with the list of available options, prompting their accession numbers, and asking you to choose a file by its number in the list, if running interactively. If you want to run `gwascat.download` automatically, and you know the number already, you can supply it using `filenum` argument, so R won't bug you.

```{r eval FALSE}
ds <- gwascat.download(29059683, hm_only = FALSE)
```
Alternatively, and for illustrative purposes, I took a random subset from this file including 100,000 SNPs from this file, which we'll use in this tutorial. Bear in mind that the final PGS won't be a *valid* model since we randomly removed most of the SNPs



The first thing to do is to check what our file looks like.

```{r}
summary(ds)
```

